### DISCUSSION TITLE
Implementing Usage Limits for Non-Subscribed Users: 5 Questions/Hour, 15/Day

### DISCUSSION THREAD

**Discussion by crissins** [microsoft/AI-Chat-App-Hack#72] [Q&A] [answered]
I am looking to implement a usage limit for non-subscribed users to encourage subscriptions. 

Specifically, I want to restrict access to 5 questions per hour and 15 questions per day for users who have not subscribed through Stripe paywall.

I've been exploring various ways to implement this functionality efficiently and securely, ensuring that users cannot bypass these restrictions easily. However, I'm keen to gather insights, advice, or experiences from the community on the best practices for implementing such features.

Thanks :)

**Comment by john0isaac** [comment, marked as answer]
Hey @crissins,
I think that what you are looking for is (RBAC) role-based access control.
Basically, you give your users different roles according to what they are able to do.
Each role has a bunch of permissions inside it that you check before fulfilling their request.

So, a user with the role "Subscribed" can have the following permissions:
- `get:questions`
- `post:questions`

A user  with the role "Registered" but not subscribed can have the following permissions:
- `get:questions`
- `post:small-questions`

Using this you can change frontend and backend behavior according to different permissions.

For example:
- New user signs up and receives the role registered.
- Once you confirm payment via stripe the user receives the role subscribed.

As for the limit on the number of questions, I think that you will need to store the number of questions the user sent somewhere and check for it before fulfilling the user request according to their permission (small-questions or questions).

There is a video on Access control as a part of the AI Chat Hack series which might also help.

[![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/64026625/6e78aafe-b513-4f34-a225-5cf3b488e30d)](https://www.youtube.com/watch?v=oahQ22oKosE)


I hope this helps!

**Reply by crissins** [reply]
Thank you veyr much this is exactly what I was looking for
